Design docs (#20)
benbovy/xdggs: design_doc.md

# XDGGS - Design document

Xarrays extension for DGGS. Technical specifiactions.

## Goals

The goal of the `xddgs` library is to facilitate working with multiple Discrete Global Grid Systems (DGGSs) via a unified, high-level and user-friendly API that is deeply integrated with [Xarray](https://xarray.dev).

Examples of common DGGS features that `xdggs` should provide or facilitate:

> - convert a DGGS from/to another grid (e.g., a DGGS, a latitude/longitude rectilinear grid, a raster grid, an unstructured mesh)

Reviewer: In the discussion on Thursday, we pretty much agreed that this doesn't need to live in `xdggs` (though as part of the work on `xdggs` we need to ensure they exist). `xesmf` and `pyresample` seem like good candidates.

Edit: maybe that's what you meant with the section in the "non-goals"?

Author: Yes, what I mean here is that `xdggs` should provide the API for doing these operations (or at least some core API that should facilitate it). `xdggs` does not necessarily need to implement the logic: if there is a well-maintained 3rd party library that does that well, it is preferable to rely on it instead of reinventing the wheel.

About `xesmf` and `pyresample`, do they handle complex cell geometries (non-rectilinear) on the sphere?

Author: The 1st sentence of the "goals" section is particularly important: *"unified, high-level and user-friendly API that is deeply integrated with Xarray"*. It is all about the API so I expect that `xdggs` would mainly consist of adapter code that extend the Xarray API and calls routines from 3rd party libs.

Reviewer: This is mostly a API design issue: does it make sense to provide a `interpolate` accessor method if we simply forward to another library? Oftentimes, using that library directly is more efficient, since at least in the case of `xesmf` it allows applying the same multiple times without a recomputation of weights. (I don't think there's a universal answer here, it depends on the library and how it works) 

> About `xesmf` and `pyresample`, do they handle complex cell geometries (non-rectilinear) on the sphere?

not sure, but the point is that if they don't we could (try to) extend them instead of implementing this ourselves. Note that those two libraries were examples, if there's a different library that does this already I wouldn't mind using that instead.

Reviewer: > It is all about the API so I expect that xdggs would mainly consist of adapter code that extend the Xarray API and calls routines from 3rd party libs.

this would certainly be very convenient, and I didn't think of alternatives, yet (just to have more options to choose from). That will take me some time, though.

Author: > This is mostly a API design issue: does it make sense to provide a interpolate accessor method if we simply forward to another library? 

Perhaps not an interpolate accessor but conversion methods in a dggs accessor yes. I think it often makes sense since we're not only regridding but we're also conforming the output to some conventional (or standard?) DGGS representation used by `xdggs` (+ building/setting an index).

It is also about discoverability. We could certainly add tutorials showing how to regrid lat/lon 2D grids to DGGS by directly calling xesmf, pyreseample, etc., and then set the DGGS index. But I find it quite conventional to have some high-level API layer in xdggs to do the regridding and conversion all at once, since it is a very common operation.

Want to create a new DGGS Dataset from another Dataset? Just type `ds.dggs.from_<tab-completion>` and pick up the conversion method that is relevant for your dataset :).

>  at least in the case of xesmf it allows applying the same multiple times without a recomputation of weights. 

I think we could arrange the `xdggs` API so that it allows reusing interpolation weights. That doesn't seem a big issue to me.

> if there's a different library that does this already I wouldn't mind using that instead.

Totally agreed, we are on the same page!

Author: What I also tried to state in the doc (non-goals section): *"Likewise, there may be many ways of resampling a grid to a DGGS ; xdggs should support the most common methods but not try to support all of them."*

While I think we should really have conversion methods for converting lat/lon 2D grid to DGGS and vice-versa that handle 80% of the use-cases, it shouldn't prevent reusing directly xesmf, pyresample, etc. directly for the more advanced use cases.

`xdggs` should facitiltate interoperability with other existing Xarray extensions (e.g., [xvec](https://github.com/xarray-contrib/xvec) for vector data or [uxarray](https://github.com/UXARRAY/uxarray) for unstructured grids).

`xdggs` should also leverage the existing implementation of various DGGSs exposed to Python via 3rd-party libraries, here below referred as "backends". Preferrably, those backends would expose DGGS functionality in an efficient way as vectorized functions (e.g., working with NumPy arrays).

`xdggs` should try to follow standards and/or conventions defined for DGGS (see below). However, we may need to depart from them for practical reasons (e.g., common practices in popular DGGS libraries that do not fit well with the proposed standards). Strict adherence to a standard is welcome but shouldn't be enforced by all means.

`xddgs` should also try to support applications in both GIS and Earth-System communities, which may each use DGGS in slightly different ways (see examples below).

> When possible, `xddgs` operations should scale to fine resolutions (millions of cells) leveraging Xarray interoperability with Dask. This might not be always possible, though. Some operations (spatial indexing) may be hard to support at scale and it shouldn't be a high development priority.

Reviewer: there's also the idea that whenever necessary (i.e. if the python implementation turns out to be too slow), slow operations need to be written in a lower-level language

Author: Yes, for low-level languages / vectorized bindings I also tried to emphasize that in the "goals" and "backends" sections.

Here this sentence is more about horizontal scaling (I can clarify that). While there are some operations that are trivial to support with Dask, other operations like spatial indexing / partitioning / shuffling / sorting are much harder to implement in a distributed fashion. That is not impossible, though, I think we could for example look at [dask-geopandas](https://github.com/geopandas/dask-geopandas). This is pretty hard work, so I wanted to write that it is probably wiser to first try supporting spatial indexing for small amounts of data and then after see how we can scale it horizontally.

## Non-Gloals

`xdggs` should focus on providing the core DGGS functionality and operations that are listed above. Higher-level operations that can be implemented by combining together those core operations are out-of-scope and should be implemented in downstream libraries. Likewise, there may be many ways of resampling a grid to a DGGS ; `xdggs` should support the most common methods but not try to support _all of them_.

`xddgs` should try not re-inventing the wheel and delegate to Xarray API when possible.

`xdggs` does not implement any particular DGGS from scratch. `xdggs` does not aim at providing _all the functionality provided by each grid_ (e.g., some functionality may be very specific to one DGGS and not supported by other DGGSs, or some functionality may not be available yet in one DGGS Python backend).

Although some DGGS may handle both the spatial and temporal domains in a joint fashion, `xdggs` focuses primarily on the spatial domain. The temporal domain is considered as orthogonal and already benefits from many core features provided by Xarray.

## Discrete Global Grid Systems

> A Discrete Global Grid System (DGGS) can be roughly defined as a partitioning or tesselation of the entire Earth's surface into a finite number of "cells" or "zones". The shape and the properties of these cells generally vary from one DGGS to another. Most DGGSs are also hierarchical, i.e., the cells are aranged on recursively on multiple levels or resolutions. Follow the links in the subsection below for a more strict and detailled definition of a DGGS.

Reviewer: ```suggestion
A Discrete Global Grid System (DGGS) can be roughly defined as a partitioning or tessellation of the entire Earth's surface into a finite number of "cells" or "zones". The shape and the properties of these cells generally vary from one DGGS to another. Most DGGSs are also hierarchical, i.e., the cells are arranged on recursively on multiple levels or resolutions. Follow the links in the subsection below for a more strict and detailed definition of a DGGS.
```

DGGSs may be used in various ways, e.g.,

> - Applications in Earth-system modelling seem to use DGGS as grids of contiguous, fixed-resolution cells covering the entire Earth's surface or a region of interest (figure 1). This makes easier the analysis of simulation outputs on large extents of the Earth's surface. DGGS may also be used as pyramid data (multiple stacked datasets at different resolutions)

Reviewer: ```suggestion
- Applications in Earth-system modelling seem to use DGGS as grids of contiguous, fixed-resolution cells covering the entire Earth's surface or a region of interest (figure 1). This makes the analysis of simulation outputs on large extents of the Earth's surface easier. DGGS may also be used as pyramid data (multiple stacked datasets at different resolutions)
```

![figure1](https://user-images.githubusercontent.com/4160723/281698490-31cb5ce8-64db-4bbf-a0d9-a8d6597bb2df.png)
Figure 1: DGGS data as contiguous cells of fixed resolution ([source](https://danlooo.github.io/DGGS.jl/))

![figure2](https://github.com/benbovy/xdggs/assets/4160723/430fd646-220a-4027-8212-1d927bb339ba)

Figure 2: Data aggreated on DGGS (H3) sparsely distributed cells of fixed resolution ([source](https://medium.com/@jesse.b.nestler/how-to-convert-h3-cell-boundaries-to-shapely-polygons-in-python-f7558add2f63)).

![image](https://github.com/benbovy/xdggs/assets/4160723/f2e4ec02-d88e-475e-9067-e93cf185923e)

Figure 3: Raster data converted as DGGS (H3) cells of mixed resolutions ([source](https://github.com/nmandery/h3ronpy)).

### Standards and Conventions

There is no released standard yet regarding DGGS. However, there is a group working on a draft of OGC API for DGGS: https://github.com/opengeospatial/ogcapi-discrete-global-grid-systems.

Another draft of DGGS specification can be found here: https://github.com/danlooo/dggs-data-spec.

> There are some discrepancies between the proposed standards and popular DGGS libraries (H3, S2, HealPIX). For example regarding the term used to define a grid unit: The two specifications above use "zone", S2/H3 use "cell" and HealPIX uses "pixel". Although in this document we use "cell", the term to choose for `xddgs` is still open for discussion.

Reviewer: ```suggestion
There are some discrepancies between the proposed standards and popular DGGS libraries (H3, S2, HealPIX). For example regarding the term used to define a grid unit: The two specifications above use "zone", S2/H3 use "cell" and HealPIX uses "pixel". Although in this document we use "cell", the term to choose for `xdggs` is still open for discussion.
```

Author: Fixed along with a few other occurrences of that typo.

- [healpy](https://github.com/healpy/healpy): Python bindings of [HealPix](https://healpix.sourceforge.io/)
  - mostly vectorized
- [rhealpixdggs-py](https://github.com/manaakiwhenua/rhealpixdggs-py): Python/Numpy implementation of rHEALPix
- [h3-py](https://github.com/uber/h3-py): "official" Python bindings of [H3](https://h3geo.org/)
  - experimental and incomplete vectorized version of H3's API (removed in the forthcoming v4 release?)
- [h3pandas](https://github.com/DahnJ/H3-Pandas): integration of h3-py (non-vectorized) with pandas and geopandas
- [h3ronpy](https://github.com/nmandery/h3ronpy): Python bindings of [h3o](https://github.com/HydroniumLabs/h3o) (Rust implementation of H3)
  - provides high-level features (conversion, etc.) working with arrow, numpy (?), pandas/geopandas and polars
- [s2geometry](https://github.com/google/s2geometry): Python bindings generated with SWIG
  - not vectorized nor very "pythonic"
  - plans to switch to pybind11 (no time frame given)
- [spherely](https://github.com/benbovy/spherely): Python bindings of S2, mostly copying shapely's API
  - provides numpy-like universal functions

>   - not yest ready for use

Reviewer: ```suggestion
  - not yet ready for use
```

  - DGGRID implements many DGGS variants!
  - DGGRID current design makes it hardly reusable from within Python in an optimal way (the dggrid wrapper communicates with DGGRID through OS processes and I/O generated files)

## Representation of DGGS Data in Xdggs

`xdggs` represents a DGGS as an Xarray Dataset or DataArray containing a 1-dimensional coordinate with cell ids as labels and with grid name, resolution & parameters (optional) as attributes. This coordinate is indexed using a custom, Xarray-compatible `DGGSIndex`.

`xdggs` does not support a Dataset or DataArray with multiple coordinates indexed with a `DGGSIndex` (only one DGGS per object is supported).

The cell ids in the 1-dimensional coordinate are all relative to the _exact same_ grid, i.e., same grid system, same grid parameter values and same grid resolution! For simplicity, `xdggs` does not support cell ids of mixed-resolutions in the same coordinate.

### DGGSIndex

`xdggs.DGGSIndex` is the base class for all Xarray DGGS-aware indexes. It inherits from `xarray.indexes.Index` and has the following specifications:

> - It encapsulates an `xarray.indexes.PandasIndex` built from cell ids so that selection and alignment by cell id is possible

Reviewer: might need to change that for hierarchical access by cell id

Reviewer: yep, that's what I meant (hence the edit).

- Select all cells that are the closest to a collection of data points
- Select all cells that intersects with or are fully contained in a polygon

This kind of selection requires spatial indexes as this can not be done with a pandas index (see [this issue](https://github.com/benbovy/xdggs/issues/16)).

If we support spatial indexing directly in `xdggs`, we can hardly reuse Xarray's `.sel()` for spatial queries as `ds.sel(cell=shapely.Polygon(...))` would look quite confusing. Perhaps better would be to align with [xvec](https://github.com/xarray-contrib/xvec) and have a separate `.dggs.query()` method.

Alternatively, we could just get away with the conversion and cell geometry extraction methods proposed above and leave spatial indexes/queries to [xvec](https://github.com/xarray-contrib/xvec).

## Operations between similar DGGS (alignment)

TODO

## Plotting

Three approaches are possible (non-mutually exclusive):

1. convert cell data into gridded or raster data (choose grid/raster resolution depending on the resolution of the rendered figure) and then reuse existing python plotting libraries (matplotlib, cartopy) maybe through xarray plotting API

> 2. convert cell data into vector data and plot the latter via, e.g., [xvec](https://github.com/xarray-contrib/xvec) or [geopandas](https://github.com/geopandas/geopandas) API

Author: Do you mean hvplot plotting vector data like https://hvplot.holoviz.org/reference/geopandas/polygons.html?

> 2. convert cell data into vector data and plot the latter via, e.g., [xvec](https://github.com/xarray-contrib/xvec) or [geopandas](https://github.com/geopandas/geopandas) API

Reviewer: I meant this: https://discourse.holoviz.org/t/dealing-high-resolution-vector-plots-xarray-holoviz/6310

> 2. convert cell data into vector data and plot the latter via, e.g., [xvec](https://github.com/xarray-contrib/xvec) or [geopandas](https://github.com/geopandas/geopandas) API

Author: Thanks for the links! Yes it looks like a similar approach to what I suggest by using Datashader with rasterization of DGGS to a dynamic resolution and extent. This could actually be applied for polygon plots: downgrade the DGGS resolution (#18), then aggregate data, then convert it to vector data, then plot it. It would be nice to have some high-level API to do that interactively instead of requiring the user doing all of those steps manually.

> 2. convert cell data into vector data and plot the latter via, e.g., [xvec](https://github.com/xarray-contrib/xvec) or [geopandas](https://github.com/geopandas/geopandas) API

Author: I rephrased the paragraph below in [5a97e2c](https://github.com/benbovy/xdggs/pull/20/commits/5a97e2c7c405bc8bc9cfdfd801b1cd330046a5ab) to reflect that comment.
